File: src/core/cmd/events/window.rs
use glam::{IVec2, Vec2};
use serde::{Deserialize, Serialize};

/// Window-related events
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "event", content = "data", rename_all = "kebab-case")]
pub enum WindowEvent {
    /// Window was created successfully
    #[serde(rename_all = "camelCase")]
    OnCreate { window_id: u32 },

    /// Window was resized
    #[serde(rename_all = "camelCase")]
    OnResize {
        window_id: u32,
        width: u32,
        height: u32,
    },

    /// Window was moved
    #[serde(rename_all = "camelCase")]
    OnMove { window_id: u32, position: IVec2 },

    /// Window close was requested by user
    #[serde(rename_all = "camelCase")]
    OnCloseRequest { window_id: u32 },

    /// Window was destroyed
    #[serde(rename_all = "camelCase")]
    OnDestroy { window_id: u32 },

    /// Window gained or lost focus
    #[serde(rename_all = "camelCase")]
    OnFocus { window_id: u32, focused: bool },

    /// Window scale factor changed
    #[serde(rename_all = "camelCase")]
    OnScaleFactorChange {
        window_id: u32,
        scale_factor: f64,
        new_width: u32,
        new_height: u32,
    },

    /// Window was occluded (completely hidden from view)
    #[serde(rename_all = "camelCase")]
    OnOcclude { window_id: u32, occluded: bool },

    /// Window redraw was requested
    #[serde(rename_all = "camelCase")]
    OnRedrawRequest { window_id: u32 },

    /// File was dropped into window
    #[serde(rename_all = "camelCase")]
    OnFileDrop {
        window_id: u32,
        path: String,
        position: Vec2,
    },

    /// File is being hovered over window
    #[serde(rename_all = "camelCase")]
    OnFileHover {
        window_id: u32,
        path: String,
        position: Vec2,
    },

    /// Hovered file left the window
    #[serde(rename_all = "camelCase")]
    OnFileHoverCancel { window_id: u32 },

    /// System theme changed
    #[serde(rename_all = "camelCase")]
    OnThemeChange { window_id: u32, dark_mode: bool },
}

File: binding/bun/src/events/window.ts
import type { IVector2 } from './common';

// MARK: Window Events

export interface WindowCreatedEvent {
  event: 'on-create';
  data: { windowId: number };
}

export interface WindowResizedEvent {
  event: 'on-resize';
  data: { windowId: number; width: number; height: number };
}

export interface WindowMovedEvent {
  event: 'on-move';
  data: { windowId: number; position: IVector2 };
}

export interface WindowCloseRequestedEvent {
  event: 'on-close-request';
  data: { windowId: number };
}

export interface WindowDestroyedEvent {
  event: 'on-destroy';
  data: { windowId: number };
}

export interface WindowFocusedEvent {
  event: 'on-focus';
  data: { windowId: number; focused: boolean };
}

export interface WindowScaleFactorChangedEvent {
  event: 'on-scale-factor-change';
  data: {
    windowId: number;
    scaleFactor: number;
    newWidth: number;
    newHeight: number;
  };
}

export interface WindowOccludedEvent {
  event: 'on-occlude';
  data: { windowId: number; occluded: boolean };
}

export interface WindowRedrawRequestedEvent {
  event: 'on-redraw-request';
  data: { windowId: number };
}

export interface WindowFileDroppedEvent {
  event: 'on-file-drop';
  data: { windowId: number; path: string };
}

export interface WindowFileHoveredEvent {
  event: 'on-file-hover';
  data: { windowId: number; path: string };
}

export interface WindowFileHoveredCancelledEvent {
  event: 'on-file-hover-cancel';
  data: { windowId: number };
}

export interface WindowThemeChangedEvent {
  event: 'on-theme-change';
  data: { windowId: number; darkMode: boolean };
}

export type WindowEvent =
  | WindowCreatedEvent
  | WindowResizedEvent
  | WindowMovedEvent
  | WindowCloseRequestedEvent
  | WindowDestroyedEvent
  | WindowFocusedEvent
  | WindowScaleFactorChangedEvent
  | WindowOccludedEvent
  | WindowRedrawRequestedEvent
  | WindowFileDroppedEvent
  | WindowFileHoveredEvent
  | WindowFileHoveredCancelledEvent
  | WindowThemeChangedEvent;
